codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java

```diff
@@ -154,6 +154,20 @@ private void encodeNonUtf8Component(CharSequence s) {
      */
     private void encodeUtf8Component(CharSequence s) {
         for (int i = 0, len = s.length(); i < len; i++) {
+            char c = s.charAt(i);
+            if (!dontNeedEncoding(c)) {
+                if (i > 0) {
+                    uriBuilder.append(s, 0, i);
+                }
+                encodeUtf8ComponentSlow(s, i);
```

Comment:
I would create a different version of `encodeUtf8ComponentSlow` that won't ask `s.length()`: on `encodeUtf8Component` we already know it so we can save to call it again.